feed: Modify the channel feed text for muted topics.

userAdityaa opened the pull request:
Fixes #31601

This pull request updates the channel feed to display different text for muted topics. The changes ensure that users can don't get confused when there is a message only in the muted topics.

When the topics are not muted (same): 
|Before|After|
|----|----|
|<img width="1440" alt="Screenshot 2024-09-15 at 9 23 47 PM" src="https://github.com/user-attachments/assets/4e32e365-945a-435e-83e6-2cac28a01c18">|<img width="1440" alt="Screenshot 2024-09-15 at 9 23 47 PM" src="https://github.com/user-attachments/assets/4e32e365-945a-435e-83e6-2cac28a01c18">|

When the topics are muted: 
|Before|After|
|----|----|
|<img width="1440" alt="Screenshot 2024-09-15 at 9 26 57 PM" src="https://github.com/user-attachments/assets/c97274e8-0d57-49fc-b0ea-be56310eff68">|<img width="1440" alt="Screenshot 2024-10-26 at 4 57 53 AM" src="https://github.com/user-attachments/assets/a1fbe9d1-c79e-4139-bf77-a5f1f7d7673b">|
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
The "Why not start the conversation?" line should be removed.

userAdityaa commented:
Hey @alya, Can you please review this one, it seems to be a better approach than the other one.

alya commented:
Do you mean #32132?

Please explain what's different about your approach / why it's better. Does this PR address all the feedback @laurynmm left on #32132?

It will also need to pass tests to be ready for review. Thanks!

userAdityaa commented:
Hey @alya, I think so all the issue as mentioned in another PR is looked here carefully. Can you please review it once. 
For the test it is failing due to some reason that I'm not able to figure out currently, but still debugging it to get some insight.

userAdityaa commented:
There is only the commit linting issue when doing tools/lint --fix 
```
tools/lint --fix web/styles/zulip.css 
gitlint   | 1: T7 Title does not match regex (^(.+:\ )?[A-Z].+\.$): "feed_settings: Modify the channel feed text for muted topics"
gitlint   | 5: B1 Line exceeds max length (189>76): "This pull request updates the channel feed to display different text for muted topics. The changes ensure that users can don't get confused when there is a message only in the muted topics."
```

userAdityaa commented:
Hey @prakhar1144, could you please check this once?
Why am I unable to see the conflicts that have occurred?

prakhar1144 commented:
@userAdityaa You fixed the merge conflicts correctly as the zulipbot mentions "removed the has conflicts label"

So, something else was merged later which resulted in a new merge conflict. Follow the same process to fix it.

timabbott commented on web/src/narrow_banner.ts:
```diff
@@ -118,20 +118,41 @@ export function pick_empty_narrow_banner(): NarrowBannerData {
+
+    const stream_id = stream_id_str !== undefined ? Number.parseInt(stream_id_str, 10) : undefined;
+
+    const stream_subscription = get_stream_subscription(stream_id!);
```
`narrow_state.stream_sub()` does this without all the extra parsing logic.

userAdityaa replied:
I missed that. Here is the updated version:
```
const stream_subscription = narrow_state.stream_sub(current_filter);
```

timabbott commented on web/styles/zulip.css:
```diff
@@ -1332,8 +1332,8 @@ div.toggle_resolve_topic_spinner .loading_indicator_spinner {
     font-weight: 400;
-    line-height: 1;
     word-wrap: break-word;
+    line-height: 1.2;
```
What impact does this have on existing banners? This might belong as its own commit with its own commit message, since it's not directly part of the feature (though may be required for it to look good), and impacts other components.

userAdityaa replied:
I have updated the code for the previous line-height.

timabbott commented on web/src/narrow_banner.ts:
```diff
@@ -109,7 +110,6 @@ function retrieve_search_query_data(): SearchData {
 export function pick_empty_narrow_banner(): NarrowBannerData {
     const default_banner = {
         title: $t({defaultMessage: "There are no messages here."}),
-        // Spectators cannot start a conversation.
```
Why did you delete this comment?

userAdityaa replied:
It was a mistake from my side. I apologise.

userAdityaa commented:
Hi @timabbott,  ready for another review! I’ve addressed your feedback and made the necessary fixes. Thank you for your input!

alya commented:
Sorry this gone got missed earlier.

@laurynmm could you please review?

laurynmm reviewed (commented):
@userAdityaa - Thanks for working on this issue!

In general, I'm wondering why you didn't follow @timabbott's suggestion [here](https://github.com/zulip/zulip/issues/31601#issuecomment-2353749842) and use the `message_list_data` structure for this? Since we want to build the empty banner narrow for the state of the current message list, I think that would be a better direction to go than looking at the channel information.

Also, your changes here impact search views (aka when there are multiple terms in the narrow filter). The issue states that we should only show this empty narrow banner for the channel feed (aka the only term in the filter is the channel).

#### Dev environment example of a search view with all topics in channel muted:
| Main | Your changes |
| --- | --- |
| ![Screenshot from 2025-02-06 13-27-14](https://github.com/user-attachments/assets/feee2c69-e02a-4591-86cd-b21dc9324c21) | ![Screenshot from 2025-02-06 13-26-41](https://github.com/user-attachments/assets/cbf721b3-7beb-45a3-a809-bdc0a819c5e4) |

Let me know if you have any questions about the general thoughts above or the individual review comments.

userAdityaa commented on web/src/narrow_banner.ts:
```diff
@@ -124,14 +124,33 @@ export function pick_empty_narrow_banner(): NarrowBannerData {
               ),
     };
+
+    const muted_banner = {
```
Address this issue and modified the code: 

```
if(narrow_state.has_message_in_muted_topic()) {     
        return {
            title: $t({
                defaultMessage:
                    "This feed is empty, because you have muted all the topics in this channel.",
            }),
        };
    }
    ```

userAdityaa commented on web/src/narrow_banner.ts:
```diff
@@ -124,14 +124,33 @@ export function pick_empty_narrow_banner(): NarrowBannerData {
     const current_filter = narrow_state.filter();
 
-    if (current_filter === undefined || current_filter.is_in_home()) {
+    if (!current_filter || current_filter.is_in_home()) {
```
It was a mistake from my side, here update it to the previous one: 
```
if (current_filter === undefined || current_filter.is_in_home()) {
        return default_banner;
}
```

userAdityaa commented on web/src/narrow_banner.ts:
```diff
@@ -124,14 +124,33 @@ export function pick_empty_narrow_banner(): NarrowBannerData {
         return default_banner;
     }
 
+    const stream_subscription = narrow_state.stream_sub(current_filter);
```
Removed that whole logic and followed the one logic @timabbott talked about.

userAdityaa commented on web/styles/zulip.css:
```diff
@@ -1327,8 +1327,8 @@ div.toggle_resolve_topic_spinner .loading_indicator_spinner {
     font-weight: 400;
-    line-height: 1;
     word-wrap: break-word;
+    line-height: 1;
```
Fixed it. 
```
.empty-feed-notice-title {
    font-size: 1.5em;
    font-weight: 400;
    line-height: 1;
    word-wrap: break-word;
}
```

userAdityaa commented:
### Summary of Changes  

- **Improved Muted Stream Handling**: Added a check for `this._items.length === 0 && this._all_items.length > 0` to correctly detect when all messages in a narrow belong to a muted stream.  
- **New Utility Function**: Introduced `has_message_in_muted_topic` in `narrow_state.ts` as a wrapper for this condition.  
- **Integration with `narrow_banner.ts`**: Now uses `has_message_in_muted_topic` to determine when to update the `defaultMessage` accordingly.

userAdityaa commented:
While doing the changes I adjusted some `function` to be `let` instead for rewiring purpose, it should not effect the functionality, if it does provide me with the feedback, I'll adjust it accordingly.

laurynmm commented:
@userAdityaa - Thanks for all your work on this! I went ahead and pushed some updates to your changes; see notes on my updates below.

@alya - I've made some changes to the empty banner text from what's specified in the issue (#31601). See the updated screenshots below to review those updates. And the notes as to my logic in those changes.

**Updates**:
- Moved the check for muted topics in a channel feed in the empty narrow logic for that case (i.e., the only term in the filter is for the "channel" operator).
- Directly look at the `message_list.current` state in the empty narrow banner logic. We know that the current message list is defined and that it is visibly empty, so we really are checking here for whether it's not empty (`!message_lists.current.empty()`).
- In the tests, instead of building a message list with muted topics, just confirm we get the expected empty narrow banner for the two expected values of `message_lists.current.empty()`.
- Adjusted the empty banner text from what's specified in the issue:
  - Shortened the main banner text.
  - Added information about viewing the muted topics via the left sidebar.
  - Moved the link to the help center to the extra information section. Note that all our other banners that have links also have the link in the extra information area and not the main banner text.

**Updated screenshots**:

| Before | After |
| --- | --- |
| Empty channel feed with muted topics | ![Screenshot from 2025-04-01 20-20-52](https://github.com/user-attachments/assets/9d74fa37-dfe3-4c5f-9df9-72fcd3965e24) |
| Same channel feed, but after clicking on "Show all topics" | ![Screenshot from 2025-04-01 20-21-11](https://github.com/user-attachments/assets/e0b83c66-9b33-47b0-b128-0842c9a55f58) |
| Empty channel feed with no messages - muted or unmuted | ![Screenshot from 2025-04-01 20-20-58](https://github.com/user-attachments/assets/871856e5-ea05-462c-a615-7e49060870fa) |

alya commented on web/src/narrow_banner.ts:
```diff
@@ -62,6 +63,22 @@ const STARRED_MESSAGES_VIEW_EMPTY_BANNER = {
+    html: $t_html(
+        {
+            defaultMessage:
+                "Select <b>Show all topics</b> in the left sidebar to see the muted topics in this channel. Learn more about muting and unmuting topics <z-link>here</z-link>.",
```
```suggestion
                "To view a muted topic, click <b>show all topics</b> in the left sidebar and select one from the list. <z-link>Learn more</z-link>",
```

alya commented:
Looks like a good direction; I left a language tweak.

alya commented:
So it's the same if you're already in the all topics view, right? I think it's fine, and not worth over-optimizing this edge case.

userAdityaa commented:
That looks great to me! I spent a lot of time working on the test for the muted topics in the channel, but I understand it’s been removed now. Regardless, I learned a lot from the process. Thank you for making the change!

laurynmm commented:
> So it's the same if you're already in the all topics view, right? I think that's fine, and not worth over-optimizing this edge case.

Yes, it's the same if you've clicked "show all topics" in the left sidebar or not. I agree that I don't think it's worth changing the message on the main message feed once they've clicked that in the left sidebar.

Is there are way to get into that state without clicking that option from the channel feed? I'm pretty sure there isn't as showing all the topics in the left sidebar doesn't change the URL for the view. Note the URL in the updated screenshot below.

---

**Updates**:
- Revised for the updated text above. See screenshot below.
  - I know we don't put a period after "Learn more" in the empty banners for the starred messages and @-mentions views, but for some reason it looks a bit odd to me here. Maybe because the informational text is so much shorter?
  - Also, do we want a comma before the "and"? E.g., "To view a muted topic, click **show all topics** in the left sidebar, and select one from the list."

**Updated screenshot**:

![Screenshot from 2025-04-04 16-37-10](https://github.com/user-attachments/assets/ab6a470a-8a96-47d8-ab02-6eb94b0485a4)

alya commented:
Yes, let's add a comma before "and". We should keep the periods consistent, and this seems fine -- it's the equivalent of a button.

laurynmm commented:
@alya - Thanks for that confirmation!

I've added the comma to the banner message, see updated screenshot below. Marking this for integration review.

**Updated screenshot**:

![Screenshot from 2025-04-10 19-38-31](https://github.com/user-attachments/assets/4a222ff3-37e1-4479-8579-2c71c23918a2)

timabbott commented:
Merged, thanks @userAdityaa and @laurynmm!
